### PR TITLE
[WIP][Prototype][CALCITE-3340] TUMBLE as a TableFunction 

### PR DESCRIPTION
**Pull request by amaliujia**
The goal of this JIRA is to generate a logical plan for the following query:
```
SELECT *
FROM TABLE(TUMBLE(
TABLE ORDERS,
INTERVAL '10' MINUTE))
```
@julianhyde 

Can you give some feedback on this WIP PR directly?

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/tools/PlannerTest.java [hunk 1]
```diff
@@ -132,6 +132,18 @@ private void checkParseAndConvert(String query,
         + "    EnumerableTableScan(table=[[hr, emps]])\n");
   }
 
+  @Test public void testParseAndConvertTableFunctionWindowing() throws Exception {
+    checkParseAndConvert(
+            "select * FROM TABLE(TUMBLE(TABLE ORDERS, INTERVAL '1' MINUTE))",
```
Comment (amaliujia) [inline core/src/test/java/org/apache/calcite/tools/PlannerTest.java]: Note that I am using `TUMBLE` as the table function.

**Comment by LiShuMing** [comment]
Is that possible that we don't change the `Parser.jj` so that use `TUMBLE/HOP` just like `UDF/Table Function`?

So as the paper says, maybe users can extend the TableFunction later.

**Comment by amaliujia** [comment]
Hi @LiShuMing, are you saying that try to use UDF for `TUMBLE` than adding it as a built-in operator?

**Comment by LiShuMing** [comment]
No, No... `TUMBLE` is a Table Function, not a simple `UDF`.

I just doubt how to be better for users to extend `TableFunctions` like `TUMBLE`/`HOP` like the paper said.

**Comment by julianhyde** [comment]
Well, technically it is both. It is a user-defined function (UDF), and it is a table function. That is, it returns a table.

Functions that return tables can be used in the FROM clause via the TABLE operator.

All UDFs (those that return tables and those that return scalar values) occupy the same namespace.

**Comment by amaliujia** [comment]
Ok If I have understood @LiShuMing's original question: I explained why changing Parser.jj in [1].

So I think changing Parser.jj is the right approach to overload `TUMBLE`. But there is any different opinion please feel free to discuss it in the email thread. 


[1]: https://lists.apache.org/thread.html/6fa619e1e4ac6c258ed58c6527271221a2bce4f220914918c6590c9c@%3Cdev.calcite.apache.org%3E

**Comment by julianhyde** [comment]
I take your point that with the limited namespace we can't have two different functions called TUMBLE.

Therefore changing the Parser.jj seems to make sense.

However, we see the future as being TUMBLE (or any user-defined table function) in the FROM clause. Therefore how about adding parser support for TUMBLE (and HOP) in the GROUP BY clause? This would be deprecated functionality, and removed at some point. The parser would be hard-wired to create calls to that function, and therefore function does not need to be called literally TUMBLE. It could be called "TUMBLE_OLD" or whatever.

**Comment by amaliujia** [comment]
@julianhyde I like your idea. So we save "TUMBLE" for one operator and use a new name for another operator, but Parser will code one so it will create an operator directly without caring about operator name. By doing so operator overloading is also avoided. I like it.

You suggest code old TUMBLE into parser(thus rename old TUMBLE operator to maybe "TUMBLE_old"), which seems to me there will be a backward compatible issue: maybe someone is already depending on "TUMBLE" as the name to lookup the old TUMBLE operator? If this is a concern it would be better to use new name for the new operator and code the new operator into Parser. And then we should start a deprecation process at some point to say only maintain old operator for 2 ~ 3 versions, and finally switch the name back. What do you think?

**Comment by julianhyde** [comment]
Don't worry about backwards compatibility in SqlStdOperatorTable. I doubt anyone is using the fields directly (other than via our parser). We'll get a better result if we don't slavishly preserve the past.

**Comment by amaliujia** [comment]
Sounds good then.

**Comment by amaliujia** [comment]
Will close this PR. The progress now is in https://github.com/apache/calcite/pull/1587.

**Comment by 34venu** [comment]
@amaliujia Is there any calcite connection to run this tumble function
Because we are running this function in windows 7 operating system through sqlline.
As sqlline.bat
sqlline>!connect jdbc:calcite: admin admin

**Pull request closed without merging**